flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierHandler.java

```diff
@@ -56,6 +56,13 @@
 	/** The timestamp as in {@link System#nanoTime()} at which the last alignment started. */
 	private long startOfAlignmentTimestamp;
 
+	/**
+	 * Cumulative counter of bytes processed during alignment. Once we complete alignment, we will
+	 * put this value into the {@link #latestBytesProcessedDuringAlignment}.
+	 */
+	private long bytesProcessedDuringAlignment;
+	private CompletableFuture<Long> latestBytesProcessedDuringAlignment = new CompletableFuture<>();
```

Comment:
Can you elaborate? Do you mean making `CheckpointMetricsBuilder` a thread safe class as a whole? I think that would be a bit worse. With futures I can get away from this. `CheckpointMetricsBuilder` is non thread safe, it can be only used from a single thread (it can be handed over to another thread of course). At the same time, the current interface clearly shows that two of the metrics are supplied asynchronously from somewhere else, with just one simple contract: they must be supplied before building the metrics.

Reply:
No, I meant adding an explicit synchonization point while filling the builder.

In handler, on finish alignment:
```
synchronized(checkpointMetrics) {
  checkpointMetrics.setBytesProcessedDuringAlignment(bytesProcessedDuringAlignment);
  checkpointMetrics.setAlignmentDurationNanos(System.nanoTime() - startOfAlignmentTimestamp);
  ...
}
allBarriersReceivedFuture.complete(null);
```

In task thread, on checkpoint ended
```
  assert allBarriersReceivedFuture.isDone();
  synchronized(checkpointMetrics) {
     checkpointMetrics.setSyncTime(...);
     ...
     checkpointMetrics.build();
  }
```

Reply:
Alternatively, have two `CheckpointMetrics(Builder)` that are thread local for the different metrics. Then have one future in handler `CompleteableFuture<CheckpointAlignmentMetrics> alignmentFuture` (subsuming `allBarriersReceivedFuture`). That's probably much easier.

Reply:
The first option that you proposed I think is really bad. It looks like you would like to have some external synchronisation, while keeping the class itself non thread safe.

Also I would strongly against making parts of the `CheckpointMetrics(Builder)` class thread safe (it's super confusing which methods are and which are not thread safe).

On the other hand, yes, I was thinking about introducing some extra pojo, that would replace two `CompletableFuture`s with one with some poko, yes, it would be probably slightly better. Is it really worth the extra effort/code? IMO not.

Reply:
Okay then leave it as is and keep that in mind as an option when adding additional metrics.